src/eve/datamodels.py

```diff
@@ -0,0 +1,1290 @@
+
+    __datamodel_fields__: ClassVar[utils.FrozenNamespace[Attribute]]
+    __datamodel_params__: ClassVar[utils.FrozenNamespace[Type]]
+    __datamodel_validators__: ClassVar[
+        Tuple[NonDataDescriptor[DataModelTp, BoundRootValidatorType], ...]
+    ]
+
+
+class GenericDataModelTp(DataModelTp, Protocol):
+    __args__: ClassVar[Tuple[Union[Type, TypeVar]]]
+    __parameters__: ClassVar[Tuple[TypeVar]]
+    __class__: ClassVar[DataModelTp]  # type: ignore[assignment]
+
+    @classmethod
+    def __class_getitem__(
+        cls: Type[GenericDataModelTp], args: Union[Type, Tuple[Type]]
+    ) -> GenericDataModelAlias:
+        ...
+
+
+ValidatorType = Callable[[DataModelTp, Attribute, T], None]
+BoundValidatorType = Callable[[Attribute, T], None]
+
+RootValidatorType = Callable[[Type[DataModelTp], DataModelTp], None]
+BoundRootValidatorType = Callable[[DataModelTp], None]
+
+
+@typing.runtime_checkable
+class TypeWithAttrValidatorTp(Protocol):
+    """Protocol for classes defining its own custom validator (when used as fields)."""
+
+    @classmethod
+    @abc.abstractmethod
+    def __type_validator__(self) -> ValidatorType:
+        raise NotImplementedError()
+
+
+class GenericDataModelAlias(typing._GenericAlias, _root=True):  # type: ignore[call-arg,name-defined]  # typing._GenericAlias not visible
+    """Custom class alias compatible with aliases created by ``typing.Generic``.
+
+    This class emulates the :class:`typing._GenericAlias` behavior, to be
+    compatible with the mechanism of the :mod:`typing` module for ``Generic``
+    types. Basically, a :class:`typing._GenericAlias` instance is a class
+    proxy which stores a reference to the original class (``__origin__``),
+    the generic type parameters (``__parameters__``) and the concrete
+    types passed at creation (``__args__``).
+
+    Both :class:`typing._GenericAlias` and this class implement a
+    ``__mro__entries__()`` method (PEP 560) and, therefore, when
+    instances of these classes are found in the list of bases of
+    a new class, they are automatically substituted by the original
+    Python class, and the new type is created as usual.
+
+    Instances of this class work exactly in the same way, but also
+    create new actual Data Model classes during the `concretization`
+    of generic models, which are stored instead of the original
+    generic models in the ``__origin__`` attribute.
+
+    The new concrete class can be accessed as usual using
+    :class:`typing.get_origin` or by using the custom :attr:`__class__`
+    shortcut provided by this class.
+
+    Examples:
+        >>> from typing import Generic, get_origin
+        >>> @datamodel
+        ... class Model(Generic[T]):
+        ...     value: T
+        >>> assert isinstance(Model[int], GenericDataModelAlias)
+        >>> assert issubclass(get_origin(Model[int]), Model)
+        >>> assert Model[int].__class__ is get_origin(Model[int])
+        >>> print(Model[int].__class__.__name__)
```

Comment:
An example to retrieve the type parameters, e.g. `T` / `int` from `Model[int]`, would be nice. In `typing._GenericAlias` this information is not exposed, but useful occasionally.

Reply:
I added a small example of the expected results for parameters/args when dealing with generic types, but that's all. The mapping from type variables to concrete types is not exposed in `GenericAlias`, there is nothing we could here if we want to keep 100%  compatibilty .

Reply:
Correct me if I'm wrong, but `_GenericAlias` actually does expose this in `__args__`. It's just that `_GenericAlias` itself is not public. I do however see that what we expose here is on par with what is exposed indirectly in `_GenericAlias` via `get_origin`.